lib/pwnlib/reg_sort.rb

```diff
@@ -0,0 +1,221 @@
+      #
+      # @return [Array]
+      #   Array of instructions, see examples for more details.
+      #
+      # @example
+      #   regs = %w(a b c d x y z)
+      #   regsort({a: 1, b: 2}, regs)
+      #   => [['mov', 'a', 1], ['mov', 'b', 2]]
+      #   regsort({a: 'b', b: 'a'}, regs, tmp: 'X')
+      #   => [['mov', 'X', 'a'], ['mov', 'a', 'b'], ['mov', 'b', 'X']]
+      #   regsort({a: 1, b: 'a'}, regs)
+      #   => [['mov', 'b', 'a'], ['mov', 'a', 1]]
+      #   regsort({a: 'b', b: 'a', c: 3}, regs)
+      #   => [['mov', 'c', 3], ['xchg', 'a', 'b']]
+      #   regsort({a: 'b', b: 'a', c: 'b'}, regs)
+      #   => [['mov', 'c', 'b'], ['xchg', 'a', 'b']]
+      #   regsort({a: 'b', b: 'a', x: 'b'}, regs, tmp: 'y', xchg: false)
+      #   => [['mov', 'x', 'b'],
+      #       ['mov', 'y', 'a'],
+      #       ['mov', 'a', 'b'],
+      #       ['mov', 'b', 'y']]
+      #   regsort({a: 'b', b: 'a', x: 'b'}, regs, tmp: 'x', xchg: false)
+      #   => ArgumentError: Cannot break dependency cycles ...
+      #   regsort({a: 'b', b: 'c', c: 'a', x: '1', y: 'z', z: 'c'}, regs)
+      #   => [['mov', 'x', '1'],
+      #       ['mov', 'y', 'z'],
+      #       ['mov', 'z', 'c'],
+      #       ['xchg', 'a', 'b'],
+      #       ['xchg', 'b', 'c']]
+      #   regsort({a: 'b', b: 'c', c: 'a', x: '1', y: 'z', z: 'c'}, regs, tmp: 'x')
+      #   => [['mov', 'x', '1'],
+      #       ['mov', 'z', 'c'],
+      #       ['mov', 'x', 'a'],
+      #       ['mov', 'a', 'b'],
+      #       ['mov', 'b', 'c'],
+      #       ['mov', 'c', 'x'],
+      #       ['mov', 'y', 'z']]
+      #   regsort({a: 'b', b: 'c', c: 'a', x: '1', y: 'z', z: 'c'}, regs, xchg: false)
+      #   => [['mov', 'x', '1'],
+      #       ['mov', 'z', 'c'],
+      #       ['mov', 'x', 'a'],
+      #       ['mov', 'a', 'b'],
+      #       ['mov', 'b', 'c'],
+      #       ['mov', 'c', 'x'],
+      #       ['mov', 'y', 'z']]
+      def regsort(in_out, all_regs, tmp: nil, xchg: true, randomize: nil)
+        # randomize = context.randomize if randomize.nil?
+
+        # TODO(david942j): stringify_keys
+        in_out = in_out.map { |k, v| [k.to_s, v] }.to_h
+        # Drop all registers which will be set to themselves.
+        # Ex. {eax: 'eax'}
+        in_out.reject! { |k, v| k == v }
+
+        # Check input
+        if (in_out.keys - all_regs).any?
+          raise ArgumentError, format('Unknown register! Know: %p.  Got: %p', all_regs, in_out)
+        end
+
+        # Collapse constant values
+        #
+        # Ex. {eax: 1, ebx: 1} => {eax: 1, ebx: 'eax'}
+        # +post_mov+ are collapsed registers, set their values in the end.
+        post_mov = in_out.group_by { |_, v| v }.values.each_with_object({}) do |list, hash|
+          val = list.first[1]
+          next if list.size == 1 || all_regs.include?(val) || val.zero?
+          list.sort!
+          list[1..-1].each do |reg, _|
+            hash[reg] = list.first.first
```

Comment:
Sort and get `first_reg, val = list.first` on the beginning, and use `first_reg` here.